### ISSUE TITLE
Returning error codes instead of strings

### ISSUE THREAD

**Issue by LunarLambda** [libsdl-org/SDL#8150]
### Status Quo

SDL currently mostly (only?) exposes errors to the user in the form of a [global/thread-local string buffer](https://github.com/libsdl-org/SDL/blob/8924d0d9277ddaf77858a5294f62848b9279880b/src/thread/SDL_thread.c#L216-L274), which can be obtained/copied from with `SDL_GetError`/`SDL_GetErrorMsg`.

Separately to this, SDL maintains a small enum `SDL_errorcode` which is a shorthand for out-of-memory and mostly RWops/IO-related errors, though this again is not exposed to the user and seems to just be for getting consistent error messages without repetition.

Most functions in SDL return -1 on failure, aside from those that return pointers, which return NULL instead.

### The Problem

This method of error handling requires allocating and managing global state, additional allocations on the user's side (in the case of `SDL_GetErrorMsg`, which has no way to determine the full buffer size needed either?) and the returned strings aren't really useful for anything other than printing/displaying as-is to the user ([which SDL_SetError already does in some cases](https://github.com/libsdl-org/SDL/blob/8924d0d9277ddaf77858a5294f62848b9279880b/src/SDL_error.c#L53-L55)).

There's also no way to differentiate an acceptable failure of an operation and an unacceptable error, unless one starts trying to *parse* the error strings, which I assume come with no real stability guarantee (and arguably, shouldn't)

Rewriting all of these functions to return an error code directly would be nice in *theory*, in that it would get rid of all of that global state, but this is likely neither feasible nor desirable.

### Proposal

Instead, the error strings could simply be turned into/replaced by status codes, and queried by the user similar to `GetLastError()` (Win32) or `errno` (Standard C) and be translated back into string form (if so desired) either by the user themself (localised, or formatted in some other manner) or by SDL itself, using functionality similar to [FormatMessage (Win32)](https://learn.microsoft.com/en-us/windows/win32/debug/retrieving-the-last-error-code) or `strerror` (Standard C). 

This would not present a huge ABI break (`SDL_GetError`/`SDL_GetErrorMsg` could be kept for compatibility and simply be implemented in terms of the new statuscode-based API, or be kept separately entirely), and would allow users to opt into much more expressive, much more standardized modes of error handling.

The added API surface is fairly small (likely no more than 3 or 4 functions, and mostly enums) and requires fairly little additional maintenance, but the gains in usability would be comparatively significant.

**Comment by LunarLambda** [comment]
See also https://github.com/libsdl-org/SDL/issues/8129 for further motivation for allowing proper error checking via status codes

**Comment by slouken** [comment]
I've considered this, and the problem is that there are zillions of unique errors possible, some of which include runtime information in the error string. You're welcome to create a proof of concept PR, I'm curious what you have in mind.

**Comment by LunarLambda** [comment]
I'll see what I can do :)

**Comment by icculus** [comment]
Almost universally, these errors are "this is totally broken and _you can't reasonably recover from it_."  In those cases, a human readable string you show to the user or write to a log file before giving up is the right approach.

Some of this is self-apparent: I failed to initialize the video subsystem, all I can do is call printf and exit, etc, but it tends to be true at most other points. For whatever reason I can't create an audio stream or the joystick we were told existed failed to open? You're either going on without it or you're terminating the program, because what else are you _really_ going to do?

As you pointed out, just about the only thing that _is_ recoverable is filesystem API failures, and those are in very specific cases (deleted a non existent file or made a directory that already existed, etc)...other things might be recoverable by the user (disk is full, etc), but it still falls under the generic "this failed and you can't do much about it" category.

That being said: something that makes enums attractive is that apps can localize them, which they can't do with arbitrary English strings. On the downside, there's lots of stuff you can't reasonable express in an enum when a specific backend fails in a very backend-specific way, and we'd be adding more enums without warning in the future.

I'm rambling a little here, sorry. My instinct here is we maybe add a thing to the filesystem API to say "did this thing already exist when we failed?" and otherwise leave it alone, but I am willing to be proven wrong on this.

**Comment by LunarLambda** [comment]
I gave it some thought.

---

SDL currently has on the order of about 1400 `SDL_SetError` calls in `src/`. Unfortunately, I'm only one human vessel and one with fairly limited time for contributions like this at that.

However, I think I have some ideas for how this could be implemented incrementally and transparently without needing to disturb large portions of the codebase at once.

As already mentioned, `SDL_Error` contains an `int code;` field which is set to 1 if an error occured, and 0 otherwise. This lets us very easily implement a catch-all:

```c
enum SDL_ErrorCode {
    SDL_ERROR_NONE = 0,
    SDL_ERROR_UNKNOWN = 1,
    // other errors in the 'global' facility, see below for explanation
};
```

I would suggest changing the `code` field from `int` to `Uint32` or something similar to give it a stable size, but that should be a trivial change given it's only used as a boolean flag currently.

---

Then, for the allocation of *actual* error codes, to make sure they're globally unique, I suggest taking a page out of Microsoft's book and using a format similar to HRESULT codes:
![grafik](https://github.com/libsdl-org/SDL/assets/38919842/188c496d-71fb-42d3-b1f5-0dee35ad875d)

In particular, using a "facility" field to categorise error codes by their respective API (such as events, video, audio, etc..). Something like a 24-bit error code and an 8-bit API code, with API code 0 reserved for 'global'/non-specific error codes.

---

`SDL_SetError`'s implementation would be moved to `SDL_SetError2` (anti-bikeshedding name), gaining a `Uint32 code` as the first parameter before the formatting string, and `SDL_SetError(...)` would become `SDL_SetError2(1 /* SDL_ERROR_UNKNOWN */, ...)`

As it stands, this would be satisfactory to me and I think many users, and retains complete backwards-compatibility with regards to the string-based error handling of SDL2. However, this means that the full error information cannot be contained solely by the error code, so it would still be necessarily complemented by the error messages.

### A slightly more full-featured proposal

<details>
Note: while I would personally love something like this for SDL, I don't currently consider actively pursuing this idea unless given explicit approval that this is something that is wanted and would receive proper support in implementing it. I include it mostly as an example of what's possible.

Taking it one step further would be to implemented a tagged union type similar to `SDL_Event` that contains error-specific information. Translating `SDL_SetError("Line too long (tried to draw %d pixels, max %d)", numpixels, MAX_PIXELS);` into something like:

```c
// the precise structure of everything is up in the air. this is mostly sketch-of-concept

// error "facility" code. Ascii 'R' (bikeshedded magic number)
#define SDL_ERROR_API_RENDER 0x52

typedef enum SDL_RenderErrorCode {
    SDL_ERROR_LINE_TOO_LONG = SDL_ERRORCODE(SDL_ERROR_API_RENDER, 123),
} SDL_RenderErrorCode;

typedef SDL_LineTooLongError {
    Uint32 code;
    int tried;
    int max;
} SDL_LineTooLongError;

typedef union SDL_ErrorInfo {
    Uint32 code;
    SDL_LineTooLongError linetoolong;
} SDL_ErrorInfo;
```

This information would be stored in the thread-local `SDL_Error` structure and would be obtainable via a function similar to `SDL_GetError`. This would increase the memory footprint somewhat, though not by much (< 32 bytes in most cases) and for errors that have an unusual amount of parameters (I saw one `SDL_SetError` call that used like a dozen formatting arguments) I would simply forgo having an info type, as it's most likely not any more useful than the formatted error message. In fact, the error info structure could simply include the result of `SDL_GetError` (the error message string) as a field in all cases, providing a catch-all.

This is the approach to type-based error handling that high-level languages like Rust prefer, and while it is significantly more effort to implement, it provides significant benefits for people wishing to use SDL from other languages, as well as making its error handling much more expressive and flexible, like being able to fully translate/localize errors or otherwise transform them (for example structured logging).
</details>

**Comment by LunarLambda** [comment]
I would be happy to work on a proof-of-concept PR that implements the above proposal on a few APIs (video and render seem like good targets), unless it is decided that this is not worth to implement library-wide, in which case I hope it could still provide some insight and ideas for error handling (in the filesystem API or elsewhere) going forwards :)

**Comment by slouken** [comment]
I frequently run into the problem in Win32 APIs where the error code is sufficiently generic ("invalid parameter") and doesn't actually help me understand what I've done wrong and how to recover from it. At least with SDL errors, the programmer can look at the error message and see exactly what happened, and trace it down to a specific line of code.

I'm inclined to leave the general system alone and expand the return codes of specific APIs that can return useful information that can be acted upon by the application. Returning error vs timeout in the wait event API is a good example of this.

FWIW, we worked around this in the RWops API by adding a status field that you could examine to see what happened, but I think API-specific return codes is a more generally useful solution.

@libsdl-org/a-team , thoughts?

**Comment by icculus** [comment]
> thoughts?

Yeah, I appreciate all the effort and thought @LunarLambda put into this, but I agree with @slouken here.
